jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java

```diff
@@ -502,12 +541,9 @@ static Chunk from(ByteBuffer byteBuffer, boolean last, Consumer<ByteBuffer> rele
          * @param retainable the Retainable this Chunk links to
          * @return a new Chunk
          * @throws IllegalArgumentException if the {@code Retainable}
-         * {@link Retainable#canRetain() cannot be retained}
          */
         static Chunk asChunk(ByteBuffer byteBuffer, boolean last, Retainable retainable)
         {
-            if (!retainable.canRetain())
```

Comment:
This is the most important spot where `canRetain` is used. We must not allow linking the lifecycle of a Chunk that potentially contains a pooled buffer to the lifecycle of non-retaining retainables like EOF or EMPTY.

While this check is a nice to have extra safeguard and the current code would work without it, it was added because that error is way too easy to make and hard to track down.